Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,846 @@
+    for(size_t k=0; k<K; k++)
+      Nodes_all[k] = nullptr;
+    Nodes_all[K-1] = new Index[mpi_env.size]; // {1, 1, ..., 1}
+    int* displs = new int[mpi_env.size]; // {0, 1, ..., size-1}
+    int* ones = new int[mpi_env.size]; // {1,1, ...}
+    for(int r=0; r<mpi_env.size; r++){
+      nNodes += Nodes_all[K-1][r] = 1;
+      displs[r] = r;
+      ones[r] = 1;
+    }
+    Base::count_g(mpi_env, Nodes_all, local_version, displs, ones); // set the lower layers' nNodes and Nodes_all[k]
+
+    allocate();
+
+    indices[0] = 0;
+    for(int r=0; r<mpi_env.size; r++){
+      indices[r+1] = indices[r] + Nodes_all[K-2][r];
+    }
+    Base::set_g(mpi_env, Nodes_all, local_version);
+    
+    for(size_t k=0; k<K; k++){
+      delete[] Nodes_all[k];
+    }
+    delete[] Nodes_all;
+    delete[] displs;
+    delete[] ones;
+  }
+
+  /*! \brief Refresh the data by MPI collective communication.
+   *
+   * The NdFlattener must have been constructed by MPI collective communication.
+   * Now refresh the values with another collective communication. The subarray lengths 
+   * resulting from both collective communications must coincide, as the indices arrays
+   * are not changed.
+   * \param[in] mpi_env - The MPI environment used for communication.
+   * \param[in] local_version - The local NdFlattener structure.
+   */
+  template<typename MPI_Environment_type>
+  void refresh( 
+    MPI_Environment_type const& mpi_env,
+    Base const* local_version 
+  ) {
+    Index* Nodes_all_0 = nullptr;
+    int* displs = new int[mpi_env.size]; // {0, 1, ..., size-1}
+    int* ones = new int[mpi_env.size]; // {1,1, ...}
+    for(int r=0; r<mpi_env.size; r++){
+      displs[r] = r;
+      ones[r] = 1;
+    }
+    LowestLayer::count_g(mpi_env, &Nodes_all_0, static_cast<LowestLayer const*>(local_version), displs, ones);
+    LowestLayer::set_g(mpi_env, &Nodes_all_0, static_cast<LowestLayer const*>(local_version));
+
+    delete[] Nodes_all_0; // allocated by count_g
+    delete[] displs;
+    delete[] ones;
+  }
+
+protected:
+  /*! \brief Count the distributed flatteners' numbers of nodes, and set nNodes.
+   *
+   * \param[in] mpi_env - MPI environment for communication
+   * \param[out] Nodes_all - [k][r] is set to number of nodes in layer (k+1), rank r.
+   * \param[in] local_version - local instance to be send to the other processes
+   * \param[in] displs - {0,1,...,size-1}
+   * \param[in] ones - {1,1,...,1}
+   */
+  template<typename MPI_Environment_type>
+  void count_g(MPI_Environment_type const& mpi_env,
+         Index** Nodes_all,
+         CurrentLayer const* local_version,
+         int const* displs, int const* ones )
+  { 
+    assert( Nodes_all[K-1]==nullptr);
+    Nodes_all[K-1] = new Index[mpi_env.size];
+    nNodes = 0;
+    // gather numbers of nodes in the current layer from all processes
+    mpi_env.MPI_Allgatherv( &(local_version->nNodes), 1, mpi_env.mpi_index, Nodes_all[K-1], ones, displs, mpi_env.mpi_index, mpi_env.comm );
```

Comment:
Why not use MPI_Allgather and avoid the boilerplate "ones" and "displs"?

Reply:
Good idea, I changed it